src/utils/document-section-utils.ts

```diff
@@ -0,0 +1,268 @@
+      style,
+      level,
+      isChecked: true,
+      children: []
+    };
+    nodes[level] = value.children;
+    nodes[level - 1].push(value);
+    return nodes;
+  }, [[]]).shift();
+
+  return structure;
+}
+
+export function flattenStructure(structure: SectionNodeStructure[]): FlatSectionNodeStructure[] {
+  const flattenedStructure: FlatSectionNodeStructure[] = [];
+  const mutatedStrucutre: SectionNodeStructure[] = JSON.parse(JSON.stringify(structure));
+
+  for (const section of mutatedStrucutre) {
+    const baseSection = { ...section };
+    delete baseSection.children;
+    const flattenedSection: FlatSectionNodeStructure = {
+      ...baseSection,
+      childrenIds: [],
+    };
+
+    if (section.children.length) {
+      for (const child of section.children) {
+        flattenedSection.childrenIds.push(child.id);
+      }
+
+      flattenedStructure.push(flattenedSection);
+      flattenedStructure.push(...flattenStructure(section.children));
+    } else {
+      delete section.children;
+      flattenedStructure.push(flattenedSection);
+    }
+  }
+
+  return flattenedStructure;
+}
+
+function addSelectedSection(
+  flatStructure: FlatSectionNodeStructure[],
+  sectionId: string,
+  isChildId = false,
+): string[] {
+  const section = flatStructure.find(section => section.id === sectionId);
+  const sectionIdsToAdd = [sectionId];
+
+  if (isChildId) {
+    toggleDisableInput(sectionId, false);
+  }
+
+  if (section.childrenIds.length) {
+    toggleAllSectionChildElements(flatStructure, sectionId, false);
+    sectionIdsToAdd.push(...getCheckedChildSection(section, flatStructure));
+  }
+
+  for (const section of flatStructure) {
+    if (section.childrenIds.includes(sectionId)) {
+      section.isChecked = true;
+      sectionIdsToAdd.push(...addSelectedSection(flatStructure, section.id, true));
+    }
+  }
+
+  return sectionIdsToAdd;
+}
+
+function getCheckedChildSection(section: FlatSectionNodeStructure, flatStructure: FlatSectionNodeStructure[]): string [] {
+  const checkedChildSection: string[] = [];
+
+  for (const id of section.childrenIds) {
+    const childSection = flatStructure.find(section => section.id === id);
+
+    if (childSection?.isChecked) {
+      checkedChildSection.push(childSection.id);
+      checkedChildSection.push(...getCheckedChildSection(childSection, flatStructure));
+    }
+  }
+
+  return checkedChildSection;
+}
+
+export function toggleAllSectionChildElements(
+  flatStructure: FlatSectionNodeStructure[],
+  sectionId: string,
+  isDisabled: boolean
+): void {
+  const section = flatStructure.find(section => section.id === sectionId);
+
+  if (section?.childrenIds.length && section?.isChecked) {
+    for (const id of section.childrenIds) {
+      toggleDisableInput(id, isDisabled);
+      toggleAllSectionChildElements(flatStructure, id, isDisabled);
+    }
+  }
+}
+
+export function filterDocumentSections(
+  renderedDoc: HTMLElement,

```

Comment:
Interacting with HTML directly is bad as it avoids type checking, and avoids encapsulation which can cause the wrong element being selected. You should manipulate the logical data, and generate the appropriate HTML that reflects that state.

Reply:
The data being manipulated comes from the document its self and is then rendered as a preview. Unlike the other toggles that current exist where we just build elements and append around the document, we have to manipulate the rendered doc in this case. I could find much of a better way to do it so if you have a better solution I'm all ears.